open-api/rest-catalog-open-api.yaml

```diff
@@ -3647,6 +3818,176 @@ components:
             type: integer
           description: "List of equality field IDs"
 
+    PreplanTableRequest:
+      type: object
+      required:
+        - table-scan-context
+      properties:
+        table-scan-context:
+          $ref: '#/components/schemas/TableScanContext'
+
+    PlanTableRequest:
+      type: object
+      required:
+        - table-scan-context
+      properties:
+        table-scan-context:
+          $ref: '#/components/schemas/TableScanContext'
+        plan-task:
+          $ref: '#/components/schemas/PlanTask'
+        stats-fields:
+          description:
+            A list of fields that the client requests the server to send statistics
+            in each `FileScanTask` returned in the response
+          type: array
+          items:
+            $ref: '#/components/schemas/FieldName'
+
+    TableScanContext:
+      anyOf:
+        - $ref: '#/components/schemas/SnapshotScanContext'
+        - $ref: '#/components/schemas/IncrementalSnapshotScanContext'
+
+    BaseTableScanContext:
+      discriminator:
+        propertyName: type
+        mapping:
+          snapshot-scan: '#/components/schemas/SnapshotScanContext'
+          incremental-snapshot-scan: '#/components/schemas/IncrementalSnapshotScanContext'
+      type: object
+      required:
+        - type
+      properties:
+        type:
+          type: string
+
+    SnapshotScanContext:
+      description: context for scanning data in a specific snapshot
+      type: object
+      allOf:
+        - $ref: '#/components/schemas/BaseTableScanContext'
+      required:
+        - type
```

Comment:
Right now, we don't require `snapshot-id`. The reason I suggested that originally was to avoid requiring an extra call to the service to load the table to get the snapshot. However, I think that not requiring it can lead to some odd behavior when the client and server are out of sync with what the latest snapshot is.

The client might thing (and could possibly log) that it is reading the latest snapshot it knows about, `s1`, and the service may know about a newer snapshot, `s2`, and plan with that instead. That could lead to a confusing situation where a user may not know what version of a table was actually read.

That kind of problem is avoided by requiring a specific snapshot ID in the request, but because snapshot IDs are unique it would require that the client has loaded the table metadata. I think that is okay and I would rather not have cases where the client and service can get out of sync. What do you think, @rahil-c, @amogh-jahagirdar, @danielcweeks ?

Reply:
@rdblue Actually think this makes alot of sense, thanks for bringing this up as I didn't think about this originally. I think the assumption I was making before was that when the client makes a loadTable request (it would get the latest table metadata) and have access to the latest snapshot. However in between the time client loads the table metadata and making the preplan or plan request Its possible that a new snapshot might have been added (the service will know about this). 

In this case I do think now that we should require a `snapshot-id` to be present at the time of the request to ensure there is no out of sync issues.

cc @jackye1995

Reply:
Will add this snapshot-id as required

Reply:
@rdblue Agreed. It makes sense to avoid scenarios where the client thinking it read snapshot 1 when in reality the server planned snapshot 2 due to a concurrent write. Forcing a user to specify a fixed snapshot avoids this problem! 

Alternatively the server could be forced to return the actually scanned snapshot ID but that just complicates the protocol/data model.